src/uct/ib/mlx5/gga/gga_mlx5.c

```diff
@@ -204,7 +203,11 @@ static int uct_ib_mlx5_gga_check_device(struct ibv_device *device)
         goto out_close_ctx;
     }
 
-    result = uct_ib_mlx5_devx_check_xgvmi(cap, ibv_get_device_name(device));
+    /*
+     * This check does not guarantee that XGVMI works with indirect mkeys.
+     * We can filter out devices that does not support XGVMI at all.
+     */
+    result = uct_ib_mlx5_devx_check_xgvmi(cap);
```

Comment:
will it guarantee?
```suggestion
    result = uct_ib_mlx5_devx_check_xgvmi(cap) &&
             uct_ib_mlx5_devx_allow_xgvmi_access(device, uct_ib_mlx5_flush_rkey_make(), 1);
```
+minor refactoring of `uct_ib_mlx5_devx_allow_xgvmi_access`

Reply:
I would love it to be that simple, but of course it doesn't fly.
We need to allow XGVMI on some real valid mkey, not a dummy one like this

Reply:
that's the difference with
```
    if ((cap_2 != NULL) && (md->flush_mr != NULL) &&
        (md->super.flush_rkey != UCT_IB_MD_INVALID_FLUSH_RKEY) &&
        uct_ib_mlx5_devx_check_xgvmi(cap_2)) {
            status = uct_ib_mlx5_devx_allow_xgvmi_access(
                                                md, md->super.flush_rkey, 1);
```
in md_open?

Reply:
The difference is that flush_rkey is a real mkey created in md_open in `uct_ib_mlx5_devx_init_flush_mr`.
I just reuse that key for XGVMI check

Reply:
Ok, I see what you mean. And of course I agree that it would be nice to detect XGVMI support before creating MD.

In order to detect whether XGVMI works with indirect mkey, we need to create either KSM or KLM mkey.
So far I found only 1 function that can create such a key: `uct_ib_mlx5_devx_reg_ksm`.
This function is called indirectly from `uct_ib_mlx5_devx_init_flush_mr`, and that's why it works for our purpose.
`uct_ib_mlx5_md_buf_alloc` uses umem, so it's not suitable for this purpose.

Now, `uct_ib_mlx5_devx_reg_ksm` depends on MD, namely it uses 2 things from it:
- `md->mkey_tag`. I think we could get rid of this dependency if needed
- `uct_ib_mlx5_devx_md_get_pdn(md)`. This one is harder, because it uses allocated `md->super.pd`

So if we want to go this way we need to refactor `uct_ib_mlx5_devx_reg_ksm`, allocate temp pd with `ibv_alloc_pd`, and then create a dummy key. So in principle this is doable, but requires a bit more refactoring

Reply:
can you pls check the case with `ucs_info -d` described in https://github.com/openucx/ucx/pull/10372 which I fixed recently? I guess this PR in current state is a kind of regression..

Reply:
I'm trying to implement it the way you proposed, it should be consistent then

Reply:
Ok, it's implemented, with a little hack to avoid massive refactoring.

Reply:
so it LGTM if uct_ib_mlx5_devx_dummy_md_init/clenup will be removed in next PR with corresponding refactoring, I think it could be useful for IB MD as well for other caps which can be detected only in runtime.

Reply:
There is another tricky issue with this approach..
I see that many fork tests are failing in CI, and the root cause is the initialization order: we must call `uct_ib_fork_init` before any `ibv_create_*` call:
```
/*
169  * ibv_fork_init testing - if fork support is requested then ibv_fork_init                                                       
170  * should be called right at the beginning of the verbs initialization flow, before ibv_create_* call.
171  *
172  * Known limitations:
173  * If ibv_fork_init is called after ibv_create_* functions - it will have no effect.
174  * OMPI initializes verbs many times during initialization in the following verbs components:
175  *      oob/ud, btl/openib, mtl/mxm, pml/yalla, oshmem/ikrit, oshmem/yoda, ompi/mca/coll/{fca,hcoll}
176  *
177  * So, ibv_fork_init should be called once, in the beginning of the init flow of every verb component
178  * to proper request fork support.
179  *
180  */
```
Violation of this rule leads to warnings in tests and potential fork errors at runtime.
So I tend to revert the change for detecting XGVMI support before MR is created: it was already a bit hack-ish, and with this fork limitation it is just not worth it